Use continue instead of pass & implement tests for `vert_coords.py`

lewisblake opened the pull request:
## Change Summary

Fixes `pass` instead of `continue` error in `vert_cords.py` and increases test coverage of `vert_coords.py` by implementing tests and removing unused code.

## Related issue number

NA

## Checklist

* [ ] Start with a draft-PR
* [x] The PR title is a good summary of the changes
* [x] PR is set to AeroTools and a tentative milestone
* [x] Documentation reflects the changes where applicable
* [x] Tests for the changes exist where applicable
* [x] Tests pass locally
* [x] Tests pass on CI
* [x] At least 1 reviewer is selected
* [x] Make PR ready to review

lewisblake committed:
Use continue instead of pass in vert_coords.py

avaldebe reviewed (commented):
is this change covered by the test suite?

lewisblake commented:
No, it has been specifically excluded from the testing coverage. This is probably not desired anymore. https://github.com/metno/pyaerocom/blob/7de51e0885a61e3a7eaa137f3a52f0ecfe981ca8/pyproject.toml#L136

lewisblake commented:
Will need to write tests for `vert_coords.py` because we using this code elsewhere now.

First step is add to test coverage.

lewisblake committed:
include vert_coords.py in testing coverage

lewisblake commented:
The code coverage indicates we are not testing this module. There is no `test_vert_coords` either.

Commit:
Add type hints to vert_coords.py

Commit:
WIP

Commit:
Implement tests vert_coords.py

thorbjoernl commented:
There seem to be a number of functions in `vert_coords.py` that aren't being used anywhere in the code base. Can these be removed? This seems to apply to:

- `is_supported()`
- `pressure2altitude()`
- ~~`geopotentialheight2altitude()` (Which is also unimplemented)~~

Commit:
Remove unused functions in

Removes the following functions which aren't used from vert_coords.py:
- `is_supported()`
- `pressure2altitude()`

Commit:
Test lev_increases_with_alt

Commit:
test `extract_1D_subset_from_data()`

Commit:
Test check_access

Commit:
Testing

Commit:
Remove commented code

Commit:
Remove unused code

Commit:
Test atmosphere_sigma_coordinate_to_pressure

Commit:
Test geopotentialheight2altitude

Commit:
Test AltitudeAccess.search_aux_coords

Commit:
Remove unused code

avaldebe commented on tests/test_vert_coords.py:
```diff
@@ -0,0 +1,125 @@
+            assert len(result) == len(ptop)
+
+
+@pytest.mark.parametrize(
+    "a,b,ps,p0,exception",
+    (
+        pytest.param(np.ones(10), np.ones(5), 1, 0, ValueError, id="test-ab-length-mismatch"),
+        pytest.param(np.ones(10), np.ones(10), 1, None, None, id="test-p0-is-none"),
+        pytest.param(np.ones(10), np.ones(10), 1, 1, None, id="test-p0-is-not-none"),
+    ),
+)
+def test_atmosphere_hybrid_sigma_pressure_coordinate_to_pressure(
+    a: np.ndarray, b: np.ndarray, ps: float, p0: float | None, exception
+):
+    if exception is not None:
+        with pytest.raises(exception):
+            atmosphere_hybrid_sigma_pressure_coordinate_to_pressure(a, b, ps, p0)
+    else:
+        result = atmosphere_hybrid_sigma_pressure_coordinate_to_pressure(a, b, ps, p0)
+
+        assert len(a) == len(b) == len(result)
```
please consider to split into 2 test functions:
- "happy path" test function: no exceptions are raised; and
- "something went wrong" test function: exceptions are expected

avaldebe commented on tests/test_vert_coords.py:
```diff
@@ -0,0 +1,125 @@
+            assert vert.lev_increases_with_alt == expected
+
+
+class TestAltitudeAccess:
+    def test_exceptions(self):
+        with pytest.raises(ValueError):
+            AltitudeAccess(None)
+
+    def test_get_altitude(self, alt: AltitudeAccess):
+        with pytest.raises(NotImplementedError):
+            alt.get_altitude(None, None)
+
+    @lustre_unavail
+    @pytest.mark.xfail
+    # TODO: This test is currently failing because of a ValueError caused by
+    # a variable name containing an underscore. This seems like an unrelated
+    # issue to me, so I have marked it xfail for now.
+    def test_check_access(self, alt: AltitudeAccess):
+        assert alt.check_altitude_access()
+
+    def test_search_aux_coords(self, alt: AltitudeAccess):
+        with pytest.raises(CoordinateNameError):
+            alt.search_aux_coords(["gkjfdshglk"])
+
+        assert alt.search_aux_coords(["lat"])
+        assert alt.search_aux_coords("lon")
+        assert not alt.search_aux_coords("z")
+        assert alt.search_aux_coords(["lat", "lon", "time"])
+        assert not alt.search_aux_coords(["lat", "lon", "time", "z"])
```
I see no need for a test class, please consider using test functions instead

Commit:
Fix code review feedback

avaldebe commented on pyaerocom/vert_coords.py:
```diff
@@ -54,14 +57,17 @@ def atmosphere_sigma_coordinate_to_pressure(sigma, ps, ptop):
+            try:
+                ptop = float(ptop)
+            except ValueError as e:
+                raise ValueError(f"Invalid input for ptop. Need floating point\nError: {repr(e)}")
```
you can simplify the error section with 
``` python
    if not isinstance(ptop, float | np.ndarray):
        try:
            ptop = float(ptop)
        except ValueError as e:
            raise ValueError("Invalid input for ptop. Need floating point") from e
```

Using `|` on `isinstance(ptop, float | np.ndarray)` is a [Python 3.10 feature](https://docs.python.org/3/library/stdtypes.html#types-union).

The mechanism behind `raise ValueError(...) from e` is called [exception chaining](https://docs.python.org/3/tutorial/errors.html#exception-chaining), and will show the original exception as part of the stack trace.

thorbjoernl replied:
Fixed.

avaldebe commented on pyaerocom/vert_coords.py:
```diff
@@ -112,14 +118,14 @@ def geopotentialheight2altitude(geopotential_height):
     Parameters
     ----------
-    geopotential_height
+    geopotential_height : float
```
annotating the return of this function should add this info to the docs,
and it will provide type hints to your IDE
``` python
def geopotentialheight2altitude(geopotential_height: float) -> float:
```

avaldebe commented on pyaerocom/vert_coords.py:
```diff
@@ -112,14 +118,14 @@ def geopotentialheight2altitude(geopotential_height):
     Computed altitude levels
     """
-
+    raise NotImplementedError("This function is a dummy function")
```
can this function be removed?

thorbjoernl replied:
It is being referenced here: https://github.com/metno/pyaerocom/blob/8e09efc1e48105b50bcf385a4f31a0ce53157ce8/pyaerocom/vert_coords.py#L206 but isn't otherwise used. If removing this doesn't break any tests I'll remove it.

thorbjoernl replied:
No tests broke from removing it, so I removed it.

avaldebe commented on tests/test_vert_coords.py:
```diff
@@ -0,0 +1,125 @@
+        vert.calc_pressure(np.ones(5))
+
+
+@pytest.mark.parametrize(
+    "name,expected,exception",
+    (
+        pytest.param("altitude", True, None, id="var-name-1"),
+        pytest.param("atmosphere_sigma_coordinate", False, None, id="var-name-2"),
+        pytest.param("atmosphere_hybrid_sigma_pressure_coordinate", False, None, id="var-name-3"),
+        pytest.param("asc", False, None, id="standard-name-1"),
+        pytest.param("z", True, None, id="standard-name-2"),
+        pytest.param("gph", None, ValueError, id="valueerror"),
+    ),
+)
+def test_VerticalCoordinate_lev_increases_with_alt(
+    name: str, expected: bool | None, exception: Exception | None
+):
+    vert = VerticalCoordinate(name)
+    if exception:
+        with pytest.raises(exception):
+            vert.lev_increases_with_alt
+    else:
+        assert vert.lev_increases_with_alt == expected
```
please put the exceptions on a separate test

thorbjoernl replied:
Done

avaldebe commented on tests/test_vert_coords.py:
```diff
@@ -0,0 +1,125 @@
+    result = atmosphere_sigma_coordinate_to_pressure(sigma, ps, ptop)
+
+    assert type(result) == type(ptop)
+    if isinstance(result, np.ndarray):
+        assert len(result) == len(ptop)
```
Is it possible to check the result on the `else`?
Also, what should be the value?
You can numerical values with `assert result == pytest.approx(expected_reesult)`

thorbjoernl replied:
I've added expected values, and an additional testcase using a sine-wave to cover a broader range of the function domain.

avaldebe commented on tests/test_vert_coords.py:
```diff
@@ -0,0 +1,125 @@
+        geopotentialheight2altitude(5)
+
+
+def test_VerticalCoordinate_exceptions():
+    # Unsupported variable name during initialization.
+    with pytest.raises(ValueError):
+        VerticalCoordinate("jtlkjhsklrg")
+
+    # Attempt calculating pressure for unsupported variable.
+    vert = VerticalCoordinate("asc")
+    with pytest.raises(CoordinateNameError):
+        vert.calc_pressure(np.ones(5))
```
please split this test or use `pytest.mark.parameterize`.
As written, a failure on the 1st assert will shadow a failure on the second.

thorbjoernl replied:
Done

avaldebe commented on tests/test_vert_coords.py:
```diff
@@ -0,0 +1,125 @@
+    assert alt.check_altitude_access()
+
+
+def test_AltitudeAccess_search_aux_coords(alt: AltitudeAccess):
+    with pytest.raises(CoordinateNameError):
+        alt.search_aux_coords(["gkjfdshglk"])
```
please put the exception part on a separate test

thorbjoernl replied:
Done.

avaldebe reviewed (changes requested):
good effort, please work a bit more on the tests

Commit:
Remove geopotentialheight function and simplify exceptions

Commit:
Test result of atmosphere_sigma_coordinate_to_pressure

Commit:
Test function output values in vert_coords.py

Commit:
Split test_lev_increases_with_alt()

Commit:
Split test_AltitudeAccess_search_aux_coords()

Commit:
Remove code

avaldebe commented on test.py:
what is the purpose of this file?
are the methods here the same as the ones on `vert_coords`?

thorbjoernl commented:
While fixing the CR issues I discovered that there were a number of references to code I had previously removed remaining in the code. These were not called in the code, and mostly used [here](https://github.com/metno/pyaerocom/blob/b664a5046853d7b467c1fb971eabbbbef373c626/pyaerocom/griddeddata.py#L1349) which is an elif clause which always raises an exception, and is thus unlikely to be part of the program flow. I've cut out most of that clause, as well as any logic related to the `vert_cord.py` check_altitude_access() without breaking any tests.

avaldebe commented:
> While fixing the CR issues I discovered that there were a number of references to code I had previously removed remaining in the code. These were not called in the code, and mostly used [here](https://github.com/metno/pyaerocom/blob/b664a5046853d7b467c1fb971eabbbbef373c626/pyaerocom/griddeddata.py#L1349) which is an elif clause which always raises an exception, and is thus unlikely to be part of the program flow. I've cut out most of that clause, as well as any logic related to the `vert_cord.py` check_altitude_access() without breaking any tests.

I do not understand what are you trying to say hare.
Where the removed, non-implemented methods, used somewhere?
Then, what did you do with the calls?
Did you removed them?

thorbjoernl commented:
It is basically cleaning up left overs from my original cleanup. I had removed some functions, but missed functions which called on the removed functions. This didn't surface because that path is only in the mentioned elif clause. This just removes the remaining dangling references. So in short, yes, I removed them.

thorbjoernl commented:
> > While fixing the CR issues I discovered that there were a number of references to code I had previously removed remaining in the code. These were not called in the code, and mostly used [here](https://github.com/metno/pyaerocom/blob/b664a5046853d7b467c1fb971eabbbbef373c626/pyaerocom/griddeddata.py#L1349) which is an elif clause which always raises an exception, and is thus unlikely to be part of the program flow. I've cut out most of that clause, as well as any logic related to the `vert_cord.py` check_altitude_access() without breaking any tests.
> 
> I do not understand what are you trying to say hare. Where the removed, non-implemented methods, used somewhere? Then, what did you do with the calls? Did you removed them?

avaldebe reviewed (approved):
🚀 🎆
